src/openforms/appointments/contrib/base.py

```diff
@@ -0,0 +1,163 @@
+        @return: List of `AppointmentProduct`
+        """
+        raise NotImplementedError()
+
+    def get_locations(self, products: list) -> list:
+        """
+        Retrieve all available locations for given `products`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @return: List of `AppointmentLocation`
+        """
+        raise NotImplementedError()
+
+    def get_dates(
+        self,
+        products: list,
+        location: AppointmentLocation,
+        start_at: date = None,
+        end_at: date = None,
+    ) -> list:
+        """
+        Retrieve all available dates for given `products` and `location`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @param location: An `AppointmentLocation`, as obtained from `get_locations`.
+        @param start_at: The start `date` to retrieve available dates for. Default: `date.today()`.
+        @param end_at: The end `date` to retrieve available dates for. Default: 14 days after `start_date`.
+        @return: List of `date`
+        """
+        raise NotImplementedError()
+
+    def get_times(
+        self, products: list, location: AppointmentLocation, day: date
+    ) -> list:
+        """
+        Retrieve all available times for given `products`, `location` and `day`.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @param location: An `AppointmentLocation`, as obtained from `get_locations`.
+        @param day: A `date` to retrieve available times for.
+        @return: List of `datetime`.
+        """
+        raise NotImplementedError()
+
+    def get_calendar(
+        self,
+        products: list,
+        location: AppointmentLocation,
+        start_at: date = None,
+        end_at: date = None,
+    ) -> OrderedDict:
+        """
+        Retrieve a calendar.
+
+        WARNING: This default implementation has significant performance issues.
+        You can override this function with a more efficient implementation if
+        the service supports it.
+
+        @param products: List of `AppointmentProduct`, as obtained from `get_available_products`.
+        @param location: An `AppointmentLocation`, as obtained from `get_locations`.
+        @param start_at: The start `date` to retrieve available dates for. Default: `date.today()`.
+        @param end_at: The end `date` to retrieve available dates for. Default: 14 days after `start_date`.
+        @return: `OrderedDict` where each key represents a date and the values is a list of times.
+        """
+        days = self.get_dates(products, location, start_at, end_at)
+
+        result = OrderedDict()
+
+        try:
+            for day in days:
+                times = self.get_times(products, location, day)
+                result[day] = times
+        except Exception as e:
+            logger.exception(e)
```

Comment:
I'm hesitant about this pokemon-style exception catching. This will also catch programming errors like `KeyError`, `ValueError`...

For the end user, the end-result is the same - they can't proceed. So I'd much rather have the error bubble all the way up and making use of proper error processing in the client side to display a notice to the user something went wrong, at least that will stop them from re-trying.

Reply:
I kept catching most errors and pretend nothing happened but made it more specific and still raise if a real unexpected error is thrown.